art/attacks/evasion/composite_adversarial_attack.py

```diff
@@ -0,0 +1,669 @@
+
+    def caa_hue(
+        self, data: "torch.Tensor", hue: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for hue component.
+        :param data: A tensor of a batch of original inputs to be attacked.
+        :param hue: Specify the hue shift angle.
+        :param labels: A tensor of a batch of the original labels to be predicted.
+        :return: The perturbed data and the corresponding hue shift angle.
+        """
+        hue = hue.detach().clone()
+        hue[self.is_attacked] = 0
+        hue.requires_grad_()
+        sur_data = data.detach().requires_grad_()
+
+        return self._comp_pgd(
+            data=sur_data, labels=labels, attack_idx=0, attack_parameter=hue, ori_is_attacked=self.is_attacked.clone()
+        )
+
+    def caa_saturation(
+        self, data: "torch.Tensor", saturation: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for saturation component.
+        :param data: A tensor of a batch of original inputs to be attacked.
+        :param saturation: Specify the saturation factor.
+        :param labels: A tensor of a batch of the original labels to be predicted.
+        :return: The perturbed data and the corresponding saturation factor.
+        """
+        saturation = saturation.detach().clone()
+        saturation[self.is_attacked] = 1
+        saturation.requires_grad_()
+        sur_data = data.detach().requires_grad_()
+
+        return self._comp_pgd(
+            data=sur_data,
+            labels=labels,
+            attack_idx=1,
+            attack_parameter=saturation,
+            ori_is_attacked=self.is_attacked.clone(),
+        )
+
+    def caa_rotation(
+        self, data: "torch.Tensor", theta: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for rotation component.
+        :param data: A tensor of a batch of original inputs to be attacked.
+        :param theta: Specify the rotation angle.
+        :param labels: A tensor of a batch of the original labels to be predicted.
+        :return: The perturbed data and the corresponding rotation angle.
+        """
+        theta = theta.detach().clone()
+        theta[self.is_attacked] = 0
+        theta.requires_grad_()
+        sur_data = data.detach().requires_grad_()
+
+        return self._comp_pgd(
+            data=sur_data, labels=labels, attack_idx=2, attack_parameter=theta, ori_is_attacked=self.is_attacked.clone()
+        )
+
+    def caa_brightness(
+        self, data: "torch.Tensor", brightness: "torch.Tensor", labels: "torch.Tensor"
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        Compute the adversarial examples for brightness component.
+        :param data: A tensor of a batch of original inputs to be attacked.
```

Comment:
We need an empty line for the automated documentation generation.
```suggestion
        Compute the adversarial examples for brightness component.

        :param data: A tensor of a batch of original inputs to be attacked.
```